tools/optaplanner-rewrite/src/main/java/org/optaplanner/rewrite/v8/AsConstraintBuilder.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2021 the original author or authors.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.optaplanner.rewrite.v8;
+
+import java.util.regex.Pattern;
+
+import org.openrewrite.ExecutionContext;
+import org.openrewrite.Recipe;
+import org.openrewrite.TreeVisitor;
+import org.openrewrite.java.JavaIsoVisitor;
+import org.openrewrite.java.JavaParser;
+import org.openrewrite.java.JavaTemplate;
+import org.openrewrite.java.MethodMatcher;
+import org.openrewrite.java.search.UsesMethod;
+import org.openrewrite.java.tree.Expression;
+import org.openrewrite.java.tree.J;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class AsConstraintBuilder extends Recipe {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(AsConstraintBuilder.class);
+
+    private static final MethodMatcher MATCHER =
+            new MethodMatcher("org.optaplanner.core.api.score.stream.ConstraintStream " +
+                    "penalize(java.lang.String, org.optaplanner.core.api.score.Score)");
```

Comment:
This was a complex story, because the AST works on a binary compatibility level (not on a source compatibility level). Normally, you wouldn't notice the difference.

But the method `penalize(String, Score)` wasn't just replaced by `penalize(Score).asConstraint(String)`, as it might feel like on a source level. In fact the `ConstraintStream.penalize(String, Score)` was replaced by `UniConstraintStream.penalize(Score)...`, Bi, Tri and Quad, which is the binary level's truth.